crates/flight_datafusion/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+#![allow(clippy::missing_errors_doc)]
+
+use arrow::array::RecordBatch;
+use async_stream::stream;
+use async_trait::async_trait;
+use flight_client::FlightClient;
+use futures::{Stream, StreamExt};
+use snafu::prelude::*;
+use sql_provider_datafusion::expr;
+use std::{any::Any, fmt, ops::Deref, pin::Pin, sync::Arc, task::Poll};
+use tokio::runtime::Handle;
+
+use arrow_flight::error::FlightError;
+use datafusion::{
+    arrow::datatypes::SchemaRef,
+    common::OwnedTableReference,
+    datasource::TableProvider,
+    error::{DataFusionError, Result as DataFusionResult},
+    execution::{context::SessionState, RecordBatchStream, TaskContext},
+    logical_expr::{Expr, TableProviderFilterPushDown, TableType},
+    physical_plan::{
+        project_schema, DisplayAs, DisplayFormatType, ExecutionPlan, SendableRecordBatchStream,
+    },
+    sql::TableReference,
+};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Unable to generate SQL: {source}"))]
+    UnableToGenerateSQL { source: expr::Error },
+
+    #[snafu(display("Unable to query FlightSQL: {source}"))]
+    Flight { source: flight_client::Error },
+
+    #[snafu(display("Unable to query FlightSQL: {source}"))]
+    ArrowFlight { source: FlightError },
+
+    #[snafu(display("Unable to retrieve schema"))]
+    NoSchema,
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub struct FlightSQLTable {
```

Comment:
The naming here is weird - the sources you are connecting to aren't FlightSQL endpoints - they are normal Flight endpoints that support SQL queries using the `Cmd` FlightDescriptors - which are two separate things.

Mitch is implementing a real `FlightSQL` provider - so I think we should rename this to `FlightTable`